sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java

```diff
@@ -68,15 +121,90 @@ public Calc copy(RelTraitSet traitSet, RelNode input, RexProgram program) {
           BeamCalcRel.class.getSimpleName(),
           pinput);
       PCollection<Row> upstream = pinput.get(0);
+      Schema outputSchema = CalciteUtils.toSchema(getRowType());
+
+      final SqlConformance conformance = SqlConformanceEnum.MYSQL_5;
+      final JavaTypeFactory typeFactory =
+          new JavaTypeFactoryImpl(BeamRelDataTypeSystem.INSTANCE) {
+            @Override
+            public Type getJavaClass(RelDataType type) {
+              if (type instanceof BasicSqlType || type instanceof IntervalSqlType) {
+                if (type.getSqlTypeName() == SqlTypeName.FLOAT) {
+                  return type.isNullable() ? Float.class : float.class;
+                }
+              }
+              return super.getJavaClass(type);
+            }
+          };
+      final BlockBuilder builder = new BlockBuilder();
+
+      final PhysType physType =
+          PhysTypeImpl.of(typeFactory, getRowType(), JavaRowFormat.ARRAY, false);
+
+      Expression input =
+          Expressions.convert_(Expressions.call(processContextParam, "element"), Row.class);
+
+      final RexBuilder rexBuilder = getCluster().getRexBuilder();
+      final RelMetadataQuery mq = RelMetadataQuery.instance();
+      final RelOptPredicateList predicates = mq.getPulledUpPredicates(getInput());
+      final RexSimplify simplify = new RexSimplify(rexBuilder, predicates, false, RexUtil.EXECUTOR);
+      final RexProgram program = BeamCalcRel.this.program.normalize(rexBuilder, simplify);
 
-      BeamSqlExpressionExecutor executor = new BeamSqlFnExecutor(BeamCalcRel.this.getProgram());
+      Expression condition =
+          RexToLixTranslator.translateCondition(
```

Comment:
Should there be some validation of what is actually in `program` or is it always safe to call `translateCondition`? What does it do?

Reply:
It should be always safe to call. It validates the program has a condition and if not emits 'true'. The code generator will optimize away 'if (true)' before generating the java code.